streams/shared/src/main/scala/zio/stream/experimental/ZSink.scala

```diff
@@ -356,29 +356,52 @@ object ZSink {
     fromPush(_ => Push.emit(z))
 
   /**
-   * A sink that effectfully folds its inputs with the provided function and initial state.
+   * A sink that effectfully folds its inputs with the provided function, termination predicate and initial state.
    */
-  def foldM[R, E, I, S](z: S)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, S] =
+  def foldM[R, E, I, S](z: S)(contFn: S => Boolean)(f: (S, I) => ZIO[R, E, S]): ZSink[R, E, I, S] =
     ZSink {
-      for {
-        state <- Ref.make(z).toManaged_
-        push = (is: Option[Chunk[I]]) =>
-          is match {
-            case None => state.get.flatMap(s => ZIO.fail(Right(s)))
-            case Some(is) =>
-              state.get
-                .flatMap(is.foldM(_)(f))
-                .flatMap(state.set)
-                .mapError(Left(_))
-          }
-      } yield push
+      ZManaged.succeed(contFn(z)).flatMap { shouldStart =>
```

Comment:
How about the simpler:
```scala
if (!contFn(z)) ZManaged.succeed(_ => Push.emit(z))
else for { ... } yield push
```
?

Reply:
I though that we want it to be lazily evaluated

Reply:
I don't think suspending that is particularly important, so let's go with the simpler version.